src/include/sof/drivers/dmic.h

```diff
@@ -315,15 +316,36 @@
 #define dmic_irq(dmic) dmic->plat_data.irq
 #define dmic_irq_name(dmic) dmic->plat_data.irq_name
 
+/* Per DAI instance component data */
+struct dmic_private_nonshared {
+	struct task dmicwork;	/* HW gain ramp update task */
+	int32_t startcount;	/* Counter in dmicwork that controls HW unmute sequence */
+	int32_t gain;		/* Gain value to be applied to HW */
+};
+
+/* Per DAI instance component data but need to be shared due to component
+ * methods happening on core 0 instead of pipeline core. (1) due to
+ * set_config, (2) due to probe.
+ */
+struct dmic_private_shared {
+	uint16_t enable[DMIC_HW_CONTROLLERS];	/* Mic 0 and 1 enable bits array for PDMx (1) */
+	uint32_t state;				/* Driver component state (2) */
+	int32_t gain_coef;			/* Gain update constant (1) */
+	int irq;				/* Interrupt number used (2) */
+};
```

Comment:
I'm not convinced this needs a separate structure. If we rename dmic_global_shared to dmic_shared then the above data can be in a per DAI instance array within dmic_shared. Also not following why the irq is per instance

Reply:
Both DMIC DAI are assigned an IRQ with a macro that takes DAI index (similarly as SSP). The IRQ happens to be the same currently in every platform but it depends on cAVS configuration. Both DMIC DAI are probed separately. I would not change this now.

I'll do the rename dmic_global_shared to dmic_shared. But access to global data may become via pointer to point to the same object. Or more nested levels. It might not be as "nice" code as this IMHO.

Reply:
@lgirdwood What you describe becomes a big mess as I interpret it. Can you please explain what code aspect is improved that way? Also by sharing to global more than necessary allows one DMIC DAI instance to mess up another.